fastestimator/trace/adapt/lr_scheduler.py

```diff
@@ -37,38 +43,176 @@ class LRScheduler(Trace):
+
+    def accumulate_single_train_loss(self, train_loss: float) -> None:
+        self.train_loss_one_cycle.append(train_loss)
+
+    def accumulate_single_eval_loss(self, eval_loss: float) -> None:
+        self.eval_loss_one_cycle.append(eval_loss)
+
+    def accumulate_all_lrs(self, lr: float) -> None:
+        self.all_train_lr.append(lr)
+
+    def gather_multiple_eval_losses(self) -> None:
+        self.all_eval_loss.append(self.eval_loss_one_cycle)
+        self.eval_loss_one_cycle = []
+
+    def gather_multiple_train_losses(self) -> None:
+        self.all_train_loss.append(self.train_loss_one_cycle)
+        self.train_loss_one_cycle = []
+
+    def lstm_stacked(self) -> tf.keras.Model:
+        model = tf.keras.Sequential()
+        model.add(tf.keras.layers.Conv1D(filters=32, kernel_size=5, activation='relu', input_shape=(300, 3)))
+        model.add(tf.keras.layers.MaxPooling1D(pool_size=2))
+        model.add(tf.keras.layers.Conv1D(filters=64, kernel_size=5, activation='relu'))
+        model.add(tf.keras.layers.MaxPooling1D(pool_size=2))
+        model.add(tf.keras.layers.LSTM(64, return_sequences=True))
+        model.add(tf.keras.layers.LSTM(64))
+        model.add(tf.keras.layers.Dense(32, activation='relu'))
+        model.add(tf.keras.layers.Dense(3, activation='softmax'))
+        return model
+
+    def predict_next_multiplier(self) -> float:
+        train_loss, missing = self._merge_list(self.all_train_loss)
+        train_loss = self._preprocess_train_loss(train_loss, missing)
+        val_loss, _ = self._merge_list(self.all_eval_loss)
+        val_loss = self._preprocess_val_loss(val_loss)
+        train_lr, _ = self._merge_list(self.all_train_lr)
+        train_lr = self._preprocess_train_lr(train_lr)
+        model_inputs = np.concatenate((train_loss, val_loss, train_lr), axis=1)
+        model_inputs = np.expand_dims(model_inputs, axis=0)
+        with tf.device("cpu:0"):
+            model_pred = feed_forward(model=self.model, x=model_inputs, training=False)
+        action = np.argmax(model_pred)
+        return self.lr_multiplier[action]
+
+    def _preprocess_val_loss(self, val_loss: list) -> np.ndarray:
+        if val_loss:
+            val_loss = self._zscore(val_loss)
+            val_loss = cv2.resize(val_loss, (1, 300), interpolation=cv2.INTER_NEAREST)
+        else:
+            val_loss = np.zeros([300, 1], dtype="float32")
+        return val_loss
+
+    def _preprocess_train_lr(self, train_lr: list) -> np.ndarray:
+        train_lr = np.array(train_lr) / train_lr[-1]
+        train_lr = cv2.resize(train_lr, (1, 300), interpolation=cv2.INTER_NEAREST)
+        return train_lr
+
+    def _preprocess_train_loss(self, train_loss: list, missing: int) -> np.ndarray:
+        target_size = (3 - missing) * 100
+        train_loss = np.array(train_loss, dtype="float32")
+        train_loss = cv2.resize(train_loss, (1, target_size))
+        train_loss = self._zscore(train_loss)
+        if train_loss.size < 300:
```

Comment:
we should fix this to actually be correct rather than continuing an accidental mistake

Reply:
the model is trained by the accidental mistake.     When we improve ARC, we will fix this.

Reply:
it may actually perform better even without retraining if we fix this

Reply:
i know, not now :)

Reply:
it's not like we're under immediate time pressure to put this specific version in

Reply:
if we retrain the model, then we have to redo the entire evaluation on multiple tasks.    The initial version should be consistent with the one we used in paper.

The ARC research will continue and we will update the model later.

Reply:
the research code will be published separately regardless. by the time the paper is out we'll already have moved on here with improvements anyways